### PR TITLE
Centralized load balancers for TaskQueue and Datastore

### PR DESCRIPTION
**Pull request by menivaitsi**

### CODE REVIEW

**Comment by menivaitsi** [comment]
Codacy is complaining about a method in TaskQueueServiceStub which comes from the SDK.

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2869/

File: AppController/djinn.rb [hunk 1]
```diff
@@ -4010,18 +4009,20 @@ def start_soap_server()
 
   def start_datastore_server
     db_master_ip = nil
+    db_proxy = nil
     verbose = @options['verbose'].downcase == 'true'
     @nodes.each { |node|
       db_master_ip = node.private_ip if node.is_db_master?
+      db_proxy = node.private_ip if node.is_shadow?
```
Comment (obino) [inline AppController/djinn.rb]: Should we use load_balancer here? and account for possibly use more than one ip?

File: AppController/djinn.rb [hunk 2]
```diff
@@ -3899,9 +3903,8 @@ def start_backup_service()
   end
 
   def start_blobstore_server()
-    # Each node has an nginx configuration to reach the datastore. Use it
-    # to make sure we are fault-tolerant.
-    BlobServer.start(my_node.private_ip, DatastoreServer::LISTEN_PORT_NO_SSL)
+    # Each node uses the active load balancer to access the Datastore.
+    BlobServer.start(get_shadow.private_ip, DatastoreServer::PROXY_PORT)
```
Comment (obino) [inline AppController/djinn.rb]: should we use load_balancer here?

File: AppServer/google/appengine/api/taskqueue/taskqueue_distributed.py [hunk 3]
```diff
@@ -99,23 +99,20 @@ def __init__(self, app_id, host, port, service_name='taskqueue'):
...
-  def _GetTQLocations(self):
-    """ Gets a list of AppScale TaskQueue servers. """
-    if os.path.exists(TASKQUEUE_LOCATION_FILE):
+  def _GetTQLocation(self):
+    """ Gets an active TaskQueue proxy. """
+    if os.path.exists(TASKQUEUE_PROXY_FILE):
       try:
-        with open(TASKQUEUE_LOCATION_FILE) as tq_file:
-          ips = [ip for ip in tq_file.read().split('\n') if ip]
+        with open(TASKQUEUE_PROXY_FILE) as tq_file:
```
Comment (obino) [inline AppServer/google/appengine/api/taskqueue/taskqueue_distributed.py]: Before we were returning a location arrays: should we still do the same to be able to hit multiple load_balancers?

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2877/

File: AppServer_Java/src/com/google/appengine/api/taskqueue/dev/AppScaleTaskQueueClient.java [hunk 4]
```diff
@@ -200,16 +199,9 @@ private String getAppId() {
...
-        String ip = "";
-        try {
-            BufferedReader br = new BufferedReader(new FileReader(TASKQUEUE_IP_FILE));
-            ip = br.readLine();
-            br.close();
-        } catch (Exception e) {
-            logger.severe("Error getting ip from taskqueue ip file: " + e.getMessage());
-        }
-        return ip;
+    private String getTaskQueueProxy() {
```
Comment (cdonati) [inline AppServer_Java/src/com/google/appengine/api/taskqueue/dev/AppScaleTaskQueueClient.java]: 👍

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2879/

File: lib/appscale_info.py [hunk 5]
```diff
@@ -61,6 +61,28 @@ def get_login_ip():
...
   return ips[0]
 
+def get_db_proxy():
+  """ Get the IP of an active DB load balancer. Since there can be
+  more than one for this deployment, we return the first one.
+
+  Returns:
+    String containing the IP of an active load balancer.
+  """
+  raw_ips = file_io.read(constants.HEAD_NODE_PRIVATE_IP_LOC)
```
Comment (obino) [inline lib/appscale_info.py]: Should this use a load-balancer IP?

File: lib/appscale_info.py [hunk 6]
```diff
@@ -61,6 +61,28 @@ def get_login_ip():
...
+  return ips[0]
+
+def get_tq_proxy():
+  """ Get the IP of an active TQ load balancer. Since there can be
+  more than one for this deployment, we return the first one.
+
+  Returns:
+    String containing the IP of an active load balancer.
+  """
+  raw_ips = file_io.read(constants.HEAD_NODE_PRIVATE_IP_LOC)
```
Comment (obino) [inline lib/appscale_info.py]: should this use a load balancer IP?

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2888/

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2890/

**Comment by menivaitsi** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2913/

**Comment by menivaitsi** [comment]
Build: https://ocd.appscale.com:8080/job/Daily%20Build/2928/

**Pull request merged**